Ractor-shareable JSON::Coder

etiennebarrie opened the pull request:
I want to be able to reuse an instance of `JSON::Coder` across Ractors. We need to add RUBY_TYPED_FROZEN_SHAREABLE to the parser config typed data definition and freeze the instance variables, after which it's possible to use a frozen coder across Ractors.

byroot commented on ext/json/ext/parser/parser.c:
```diff
@@ -1563,7 +1568,7 @@ static const rb_data_type_t JSON_ParserConfig_type = {
     },
     0, 0,
-    RUBY_TYPED_FREE_IMMEDIATELY | RUBY_TYPED_WB_PROTECTED,
+    RUBY_TYPED_FREE_IMMEDIATELY | RUBY_TYPED_WB_PROTECTED | RUBY_TYPED_FROZEN_SHAREABLE,
```
> BTW `bool parsing_name;` seems unused.

Good catch: ab5efca015167684d696980b6cec7848551ca07f

etiennebarrie replied:
@eregon Thanks for the review! I added a call to `rb_check_frozen()` in `cParserConfig_initialize`.

```console
>> JSON::Coder.new.freeze.instance_variable_get(:@parser_config).send(:initialize, {})
(irb):1:in 'JSON::Ext::ParserConfig#initialize': can't modify frozen JSON::Ext::ParserConfig: #<JSON::Ext::ParserConfig:0x0000000125000d40> (FrozenError)
```

I wasn't sure whether to add a test for that, it looked like this:

```ruby
    coder = JSON::Coder.new.freeze
    parser_config = coder.instance_variable_get(:@parser_config)
    assert_instance_of JSON::Parser::Config, parser_config
    assert_raise FrozenError do
      parser_config.send(:initialize, {})
    end
```

BTW I just realized that the `Generator::State` object is already frozen in `initialize`, I'll actually do the same for the `Parser::Config`.

byroot commented:
> I can't repro the errors with JRuby

Weird, I can't either.

etiennebarrie commented:
Yeah at first I didn't have the test at all but then when I found out that the existing `State#freeze` was subtly incorrect, I felt like it was useful. But the important test is the high-level ractor test. I'll clean that up.

etiennebarrie commented:
I've added back the test for `JSON::Parser::Config` and skipped it on JRuby.
I also moved the duplicated macro for `RUBY_TYPED_FROZEN_SHAREABLE` compatibility to `ext/json.h`.
